api/v1alpha1/zz_conversion.go

```diff
@@ -0,0 +1,583 @@
+		json.SerializerOptions{Strict: false},
+	)
+	v1beta1GroupVersion = schema.GroupVersion{Group: GroupVersion.Group, Version: "v1beta1"}
+)
+
+func init() {
+	v1beta1.AddToScheme(localScheme)
+	configv1.AddToScheme(localScheme)
+	clientgoscheme.AddToScheme(localScheme)
+	AddToScheme(localScheme)
+}
+
+// HostedCluster conversion
+func (h *HostedCluster) ConvertTo(rawDst conversion.Hub) error {
+	temp := h.DeepCopy()
+	if err := fixupHostedClusterBeforeConversion(temp); err != nil {
+		return err
+	}
+	return serializationConvert(temp, rawDst)
+}
+
+func (h *HostedCluster) ConvertFrom(rawSrc conversion.Hub) error {
+	err := serializationConvert(rawSrc, h)
+	if err != nil {
+		return err
+	}
+	return fixupHostedClusterAfterConversion(h)
+}
+
+func (n *NodePool) ConvertTo(rawDst conversion.Hub) error {
+	if n.Spec.NodeCount != nil && n.Spec.Replicas == nil {
+		n.Spec.Replicas = n.Spec.NodeCount
+	}
+	return serializationConvert(n, rawDst)
+}
+
+func (n *NodePool) ConvertFrom(rawSrc conversion.Hub) error {
+	return serializationConvert(rawSrc, n)
+}
+
+func (e *AWSEndpointService) ConvertTo(rawDst conversion.Hub) error {
+	return serializationConvert(e, rawDst)
+}
+func (e *AWSEndpointService) ConvertFrom(rawSrc conversion.Hub) error {
+	return serializationConvert(rawSrc, e)
+}
+
+func (h *HostedControlPlane) ConvertTo(rawDst conversion.Hub) error {
+	temp := h.DeepCopy()
+	if err := fixupHostedControlPlaneBeforeConversion(temp); err != nil {
+		return err
+	}
+	return serializationConvert(temp, rawDst)
+}
+
+func (h *HostedControlPlane) ConvertFrom(rawSrc conversion.Hub) error {
+	err := serializationConvert(rawSrc, h)
+	if err != nil {
+		return err
+	}
+	return fixupHostedControlPlaneAfterConversion(h)
+}
+
+func serializationConvert(from runtime.Object, to runtime.Object) error {
+	b := &bytes.Buffer{}
+	from.GetObjectKind().SetGroupVersionKind(schema.GroupVersionKind{})
+	if err := serializer.Encode(from, b); err != nil {
+		return fmt.Errorf("cannot serialize %T: %w", from, err)
+	}
+	if _, _, err := serializer.Decode(b.Bytes(), nil, to); err != nil {
+		return fmt.Errorf("cannot decode %T: %w", to, err)
+	}
+	gvks, _, err := localScheme.ObjectKinds(to)
+	if err != nil || len(gvks) == 0 {
+		return fmt.Errorf("cannot get gvk for %T: %w", to, err)
+	}
+	to.GetObjectKind().SetGroupVersionKind(gvks[0])
+	return nil
+}
+
+func fixupHostedClusterBeforeConversion(hc *HostedCluster) error {
+	if hc.Spec.Platform.AWS != nil {
+		reconcileDeprecatedAWSRoles(hc.Spec.Platform.AWS)
+	}
+	if err := reconcileDeprecatedGlobalConfig(hc.Spec.Configuration); err != nil {
+		return err
+	}
+	if err := reconcileDeprecatedNetworkSettings(&hc.Spec.Networking); err != nil {
+		return err
+	}
+	return nil
+}
+
+func fixupHostedClusterAfterConversion(hc *HostedCluster) error {
+	if hc.Spec.Platform.AWS != nil {
+		populatedDeprecatedAWSRoles(hc.Spec.Platform.AWS)
+	}
+	populateDeprecatedNetworkingFields(&hc.Spec.Networking)
```

Comment:
why do we need to populate deprecated fields? v1alpha1 has all the newest ones right?